README
NAME
    Net::LDAP::SimpleServer - Minimal-configuration, read-only LDAP server

SYNOPSIS
    WORK IN PROGRESS!! NOT READY TO USE YET!!

        package MyServer;

        use Net::LDAP::SimpleServer;

        # Or, specifying a Net::Server personality
        use Net::LDAP::SimpleServer 'PreFork';

        # using default configuration file
        my $server = Net::LDAP::SimpleServer->new();

        # passing a specific configuration file
        my $server = Net::LDAP::SimpleServer->new({
            conf_file => '/etc/ldapconfig.conf'
        });

        # passing configurations in a hash
        my $server = Net::LDAP::SimpleServer->new({
            port => 5000,
            ldap_data => '/path/to/data.ldif',
        });

        # make it spin
        $server->run();

    The default configuration file is:

        ${HOME}/.ldapsimpleserver/config

DESCRIPTION
    As the name suggests, this module aims to implement a simple LDAP
    server, using many components already available in CPAN. It can be used
    for prototyping and/or development purposes. This is NOT intended to be
    a production-grade server, altough some brave souls in small offices
    might use it as such.

    As of November 2010, the server will simply load a LDIF file and serve
    its contents through the LDAP protocol. Many operations are NOT
    available yet, notably writing into the directory tree, but we would
    like to implement that in a near future.

CONSTRUCTOR
    The constructors will follow the rules defined by Net::Server, but most
    notably we have the two forms below:

    new()
        Attempts to create a server by using the default configuration file,
        "${HOME}/.ldapsimpleserver/config".

    new( HASHREF )
        Attempts to create a server by using the options specified in a hash
        reference rather than reading them from a configuration file.

METHODS
    options()
        As specified in Net::Server, this method creates new options for
        the, server, namely:

            ldap_data - the LDIF data file used by LDIFStore

            root_dn - the administrator DN of the repository

            root_pw - the password for root_dn

    default_values()
        As specified in Net::Server, this method provides default values for
        a number of options. In Net::LDAP::SimpleServer, this method is
        defined as:

            sub default_values {
                return {
                    host         => '*',
                    port         => 389,
                    proto        => 'tcp',
                    root_dn      => 'cn=root',
                    root_pw      => 'ldappw',
                    syslog_ident => 'Net::LDAP::SimpleServer-'
                        . $Net::LDAP::SimpleServer::VERSION,
                    conf_file => DEFAULT_CONFIG_FILE,
                };
            }

        Notice that we do set a default password for the "cn=root" DN. This
        allows for out-of-the-box testing, but make sure you change the
        password when putting this to production use.

    post_configure_hook()
        Method specified by Net::Server to validate the passed options

    process_request()
        Method specified by Net::Server to actually handle one connection.
        In this module it basically delegates the processing to
        Net::LDAP::SimpleServer::ProtocolHandler.

DIAGNOSTICS
    "Error message here, perhaps with %s placeholders"
        [Description of error here]

    "Another error message here"
        [Description of error here]

        [Et cetera, et cetera]

CONFIGURATION AND ENVIRONMENT
    Net::LDAP::SimpleServer may use a configuration file to specify the
    server settings. If no file is specified and options are not passed in a
    hash, this module will look for a default configuration file named
    "${HOME}/.ldapsimpleserver/config".

        ldap_data /path/to/a/ldif/file.ldif
        #port 389
        #root_dn cn=root
        #root_pw somepassword
        #objectclass_req (true|false)
        #user_tree dc=some,dc=subtree,dc=com
        #user_id_attr uid
        #user_pw_attr password

DEPENDENCIES
    Net::LDAP

    Net::LDAP::Server

    Net::Server

    UNIVERSAL::isa

    Carp

    File::HomeDir

    File::Spec::Functions

    Scalar::Util

    Config::General

    Net::LDAP::SimpleServer::LDIFStore

INCOMPATIBILITIES
    None reported.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests to
    "bug-net-ldap-simpleserver@rt.cpan.org", or through the web interface at
    <http://rt.cpan.org>.

AUTHOR
    Alexei Znamensky "<russoz@cpan.org>"

LICENCE AND COPYRIGHT
    Copyright (c) 2010, Alexei Znamensky "<russoz@cpan.org>". All rights
    reserved.

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. See perlartistic.

DISCLAIMER OF WARRANTY
    BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
    FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
    OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
    PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
    EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
    ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
    YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
    NECESSARY SERVICING, REPAIR, OR CORRECTION.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
    REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE LIABLE
    TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
    SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
    RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
    FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
    SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
    DAMAGES.



Changes
Revision history for Net-LDAP-SimpleServer

0.0.12  [2010.11.15 15:41:39] Net::LDAP::SimpleServer DIST VERSION 0.0.12
======= ======================================================================
f305dca Bumped version to 0.0.12
9d1b1a4 Fixed root_id => root_dn in POD docs

0.0.11  [2010.11.14 11:43:03] Net::LDAP::SimpleServer DIST VERSION 0.0.11
======= ======================================================================
ab18ca8 some debug support + one more test
801b930 minor cosmetic change
9c5b6d3 t/03.param.t revamped
dbded61 added common::sense and IO::Pipe
de7f2c7 Added configuration for port 20389, to allow non-root tests
b67da4d Bumped version to 0.0.11
2ea4161 added t/04.bind.t
17db48c Fixed root_id => root_dn in POD docs

0.0.10  [2010.11.08 19:33:29] Net::LDAP::SimpleServer DIST VERSION 0.0.10
======= ==============================================================================
ef0f769 if needed creates directory for log_file and pid_file
7a19809 Multiple changes
19a76e3 added min version to Net::Server, added File::Path
09a48a2 better param checking in the constructors
8708fa5 added some tests for SimpleServer constructor
a62ac40 added a test for empty params on constructor
82cfee2 added test for non-existent ldif file as well
cb865fb new example files
0dc7aac new exaqmple and test files
a9ebf2b checking for (and sorting) MANIFEST file
62a774c minor cosmetic change
e7d8624 bumped version to 0.0.10
3c54751 improved mkdist
c6aa48c mkdist: format of Changes file

0.0.9   06.11.2010 Net::LDAP::SimpleServer DIST VERSION 0.0.9
======= ======================================================================
ed4fe82 added files to iand sorted MANIFEST
b06bb66 test for building a LDIFStore
8bd47ff LDIF file for testing
86ff9d5 test for building a ProtocolHandler
e24482c bumped version up to 0.0.9
e85df80 bumped version up to 0.0.9
fb49924 added examples/test1.ldif
07ea91e leaner message in Changes
ad77070 perltidy
95acc7e Perltidy + bump version to 0.0.7
0e49cff Bumper version to 0.0.8, removed version from perldoc
b2844fe Tidying and Testing

0.0.7   31.10.2010 Net::LDAP::SimpleServer DIST VERSION 0.0.7
======= ======================================================================
b03b9b9 tidying all *.pm and *.t files
6823852 Docs and glitches in SimpleServer.pm
48e3cd5 supressed unitialized values warning
70fa566 Added the ProtocolHandler class and its first test
d3405b2 Improved legibility and rewrote _open_ldif()
f427cc9 Added the ProtocolHandler class and its first test
ae3a4ad Proper documentation for ProtocolHandler
284d4d0 Merge branch 'master' of https://github.com/russoz/Net-LDAP-SimpleServer
9489d1a Started working on the Net::Server functions
ef87fb3 Removed dependency on the Net::LDAP::Server 1-param constructor
47ab728 updated dependency list
143c304 improved perlcritic.t
e6033ed added dependencies documentation

0.0.6   14.10.2010 Net::LDAP::SimpleServer DIST VERSION 0.0.6
======= ======================================================================
b9c1872 Requires Net::LDAP, duh

0.0.5   14.10.2010 Net::LDAP::SimpleServer DIST VERSION 0.0.5
======= ======================================================================
ca8f5d7 improved documentation
e7f6fcd adjusted documentation to/fro code
b665d11 fixed missing indirection operator
04cb3b8 removed Makefile.PL from MANIFEST
1d3d5e2 Coded configuration checking
9bfcf15 updated version number to 0.0.5
6e03cb1 changed the default configuration file
28c4cc6 Fixed documentation.
dc4dcbd added git tagging to the script
bbf4820 removed Makefile.PL

0.0.4   14.10.2010 Net::LDAP::SimpleServer DIST VERSION 0.0.4
======= ======================================================================
cb4769e moved to version 0.0.4
6471508 removed unnecessary export from Scalar::Util
90c0fab Actually implemented the store
3621005 added more documentation

0.0.3   12.10.2010 Net::LDAP::SimpleServer DIST VERSION 0.0.3
======= ======================================================================
7359e48 mkdist eye-candy on the output
f6dfd34 run through perltidy
1370f6d Multiple changes
4fadc8c More perltidy!
da21aa4 Module has been tidied up
3feaa31 Made it more generic towards module name
0ef415d Basic parameter testing is in place.
0eb50e8 initial commit
0c42e06 initial commit

0.0.1  Fri Mar 26 00:47:26 2010
       Initial release.



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name       => 'Net::LDAP::SimpleServer',
    license           => 'perl',
    dist_author       => 'Alexei Znamensky <russoz@cpan.org>',
    dist_version_from => 'lib/Net/LDAP/SimpleServer.pm',
    requires          => {
        'Net::Server'       => 0.91,     # must have a new() method
        'Net::LDAP'         => 0.4001,
        'Net::LDAP::Server' => 0.43,     # must support 2-params constructor
        'UNIVERSAL::isa'    => 0,
        'Scalar::Util'      => 0,
        'File::Path'        => 0,
        'File::Spec'        => 0,
        'File::HomeDir'     => 0,
        'File::Basename'    => 0,
        'Test::More'        => 0,
        'IO::Pipe'          => 0,
        'common::sense'     => 0,
        'version'           => 0,
        'Carp'              => 0,
    },
    add_to_cleanup => ['Net-LDAP-SimpleServer-*'],
);

$builder->create_build_script();


lib/Net/LDAP/SimpleServer/LDIFStore.pm
package Net::LDAP::SimpleServer::LDIFStore;

use strict;
use warnings;
use diagnostics;

use Carp;
use UNIVERSAL::isa;
use Scalar::Util qw(blessed reftype);
use Net::LDAP::LDIF;

use version; our $VERSION = qv('0.0.12');

sub new {
    my ( $class, $param ) = @_;
    my $self = bless( { list => undef }, $class );

    croak 'Must pass parameter' unless $param;

    $self->load($param);

    return $self;
}

sub load {
    my ( $self, $param ) = @_;

    croak 'Must pass parameter!' unless $param;

    $self->{ldifobj} = _open_ldif($param);
    $self->{list}    = _load_ldif( $self->{ldifobj} );
}

sub ldif {
    my $self = shift;
    return $self->{ldifobj};
}

#
# opens a filename, a file-handle, or a Net::LDAP::LDIF object
#
sub _open_ldif {
    my $param = shift // '';

    return $param if blessed($param) && $param->isa('Net::LDAP::LDIF');

    my $reftype = reftype($param) // '';
    if ( $reftype eq 'HASH' ) {
        croak q{Hash parameter must contain a "ldif" parameter}
          unless exists $param->{ldif};

        return Net::LDAP::LDIF->new(
            $param->{ldif},
            'r',
            (
                exists $param->{ldif_options}
                ? %{ $param->{ldif_options} }
                : undef
            )
        );
    }

    # Then, it must be a filename
    croak q{Cannot find file "} . $param . q{"} unless -r $param;

    return Net::LDAP::LDIF->new($param);
}

#
# loads a LDIF file
#
sub _load_ldif {
    my $ldif = shift;

    my @list = ();
    while ( not $ldif->eof() ) {
        my $entry = $ldif->read_entry();
        if ( $ldif->error() ) {
            print STDERR "Error msg: ",    $ldif->error(),       "\n";
            print STDERR "Error lines:\n", $ldif->error_lines(), "\n";
            next;
        }

        push @list, $entry;
    }
    $ldif->done();

    my @sortedlist = sort { uc( $a->dn() ) cmp uc( $b->dn() ) } @list;

    return \@sortedlist;
}

sub filter {
    my ( $self, $sub ) = @_;
    return () unless $self->{list};

    my @list = @{ $self->{list} };

    foreach my $index ( 0 .. $#list ) {
        delete $list[$index] unless $sub->( $list[$index] );
    }
    return @list;
}

1;    # Magic true value required at end of module
__END__

=head1 NAME

Net::LDAP::SimpleServer::LDIFStore - Data store to support C<Net::LDAP::SimpleServer>

=head1 SYNOPSIS

    use Net::LDAP::SimpleServer::LDIFStore;

    my $store = Net::LDAP::SimpleServer::LDIFStore->new();
    $store->load( "data.ldif" );

    my $store =
      Net::LDAP::SimpleServer::LDIFStore->new({ ldif => 'data.ldif' });

    my $ldif = Net::LDAP::LDIF->new( "file.ldif" );
    my $store = Net::LDAP::SimpleServer::LDIFStore->new($ldif);

    my $result = $store->filter( sub {
      $a = shift;
      $a->get_value('cn') =~ m/joe/i;
    } );

=head1 DESCRIPTION

This module provides an interface between Net::LDAP::SimpleServer and a
LDIF file where the data is stored.

As of now, this interface is quite simple, and so is the underlying data
structure, but this can be easily improved in the future.

=head1 CONSTRUCTOR 

=over

=item new()

Creates a store with no data in it. It cannot be really used like that, you 
B<must> C<< load() >> some data in it first.

=item new( FILE )

Create the data store by reading FILE, which may be the name of a file or an
already open filehandle. It is passed directly to
L<<  Net::LDAP::LDIF >>.

Constructor. Expects either: a filename, a file handle, a hash reference or
a reference to a C<Net::LDAP::LDIF> object.

=item new( HASHREF )

Create the data store using the parameters in HASHREF. The associative-array
referenced by HASHREF B<must> contain a key named C<< ldif >>, which must
point to either a filename or a file handle, and it B<may> contain a key named
C<< ldif_options >>, which may contain optional parameters used in the
3-parameter version of the C<Net::LDAP::LDIF> constructor. The LDIF file will
be used for reading the data.

=item new( LDIF )

Uses an existing C<< Net::LDAP::LDIF >> as the source for the directory data.

=back

=head1 METHODS

=over

=item load( PARAM )

Loads data from a source specified by PARAM. The argument may be in any of the
forms accepted by the constructor, except that it B<must> be specified.

=item ldif()

Returns the underlying C<< Net::LDAP::LDIF >> object.

=item filter( SUBREF )

Returns a list of entries for which the function referenced by SUBREF returns
C<true>.

=back

=for head1 DIAGNOSTICS

=for author to fill in:
    List every single error and warning message that the module can
    generate (even the ones that will "never happen"), with a full
    explanation of each problem, one or more likely causes, and any
    suggested remedies.
    =over
    =item C<< Error message here, perhaps with %s placeholders >>
    [Description of error here]
    =item C<< Another error message here >>
    [Description of error here]
    [Et cetera, et cetera]
    =back

=head1 CONFIGURATION AND ENVIRONMENT

=for author to fill in:
    A full explanation of any configuration system(s) used by the
    module, including the names and locations of any configuration
    files, and the meaning of any environment variables or properties
    that can be set. These descriptions must also include details of any
    configuration language used.
  
Net::LDAP::SimpleServer requires no configuration files or environment variables.

=head1 DEPENDENCIES

L<< UNIVERSAL::isa >>

L<< Scalar::Util >>

L<< Net::LDAP::LDIF >>

=head1 INCOMPATIBILITIES

None reported.

=head1 BUGS AND LIMITATIONS

This store does not yet support writing to a LDIF file, which makes the 
C<< Net::LDAP::SimpleServer >> a read-only server.

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-net-ldap-simpleserver@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.

=head1 AUTHOR

Alexei Znamensky  C<< <russoz@cpan.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2010, Alexei Znamensky C<< <russoz@cpan.org> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.



lib/Net/LDAP/SimpleServer.pm
package Net::LDAP::SimpleServer;

use Carp;
use common::sense;

use version; our $VERSION = qv('0.0.12');
our $personality = undef;

sub import {
    my $pkg = shift;
    $personality = shift || 'Fork';

    use Net::Server;
    eval "use base qw{Net::Server::$personality}";
    croak $@ if $@;

    @Net::LDAP::SimpleServer::ISA = qw(Net::Server);

    #use Data::Dumper;
    #print STDERR Data::Dumper->Dump( [ \@Net::LDAP::SimpleServer::ISA ],
    #    ['ISA'] );
}

use File::Basename;
use File::HomeDir;
use File::Spec;
use File::Path qw{make_path};
use Scalar::Util qw{reftype};
use Net::LDAP::SimpleServer::LDIFStore;
use Net::LDAP::SimpleServer::ProtocolHandler;

use constant BASEDIR => File::Spec->catfile( home(),  '.ldapsimpleserver' );
use constant LOGDIR  => File::Spec->catfile( BASEDIR, 'log' );
use constant DEFAULT_CONFIG_FILE => File::Spec->catfile( BASEDIR, 'config' );
use constant DEFAULT_DATA_FILE => File::Spec->catfile( BASEDIR, 'server.ldif' );

make_path(LOGDIR);

my $_add_option = sub {
    my ( $template, $prop, $opt, $initial ) = @_;

    $prop->{$opt}     = $initial;
    $template->{$opt} = \$prop->{$opt};
};

sub options {
    my ( $self, $template ) = @_;
    ### setup options in the parent classes
    $self->SUPER::options($template);

    ### add a single value option
    my $prop = $self->{server};
    $_add_option->( $template, $prop, 'ldap_data', undef );
    $_add_option->( $template, $prop, 'root_dn',   undef );
    $_add_option->( $template, $prop, 'root_pw',   undef );

    #use Data::Dumper;
    #print STDERR Data::Dumper->Dump( [$self], ['options_END'] );
}

sub default_values {
    my $self = @_;

    my $v = {};
    $v->{port}     = 389;
    $v->{root_dn}  = 'cn=root';
    $v->{root_pw}  = 'ldappw';
    $v->{log_file} = File::Spec->catfile( LOGDIR, 'server.log' );

    #$v->{pid_file} = File::Spec->catfile( LOGDIR, 'server.pid' );
    $v->{conf_file} = DEFAULT_CONFIG_FILE if -r DEFAULT_CONFIG_FILE;
    $v->{ldap_data} = DEFAULT_DATA_FILE   if -r DEFAULT_DATA_FILE;
    $v->{syslog_ident} =
      'Net::LDAP::SimpleServer-' . $Net::LDAP::SimpleServer::VERSION;
    return $v;
}

sub _make_dir {
    my $file = shift;
    return unless $file;

    my $dir = dirname($file);
    return unless $dir;
    return if -d $dir;

    make_path($dir);
}

sub post_configure_hook {
    my $self = shift;
    my $prop = $self->{server};

    #use Data::Dumper;
    #print STDERR '# ' . Data::Dumper->Dump( [$self], ['post_configure_hook'] );
    croak q{Cannot find conf file "} . $self->{server}->{conf_file} . q{"}
      if $self->{server}->{conf_file} and not -r $self->{server}->{conf_file};
    _make_dir( $self->{server}->{log_file} );
    _make_dir( $self->{server}->{pid_file} );
    croak q{Configuration has no "ldap_data" file!}
      unless exists $prop->{ldap_data};
    croak qq{Cannot read ldap_data file "} . $prop->{ldap_data} . q{"}
      unless -r $prop->{ldap_data};

    $prop->{store} =
         Net::LDAP::SimpleServer::LDIFStore->new( $prop->{ldap_data} )
      || croak q{Cannot create data store!};
}

sub process_request {
    my $self = shift;

    my $in  = *STDIN{IO};
    my $out = *STDOUT{IO};
    my $params =
      { map { ( $_ => $self->{server}->{$_} ) } qw/store root_dn root_pw/ };
    my $handler =
      Net::LDAP::SimpleServer::ProtocolHandler->new( $params, $in, $out );

    until ( $handler->handle ) {

        # empty loop
    }
    return;
}

1;    # Magic true value required at end of module
__END__

=head1 NAME

Net::LDAP::SimpleServer - Minimal-configuration, read-only LDAP server

=head1 SYNOPSIS

B<< WORK IN PROGRESS!! NOT READY TO USE YET!! >>

    package MyServer;

    use Net::LDAP::SimpleServer;

    # Or, specifying a Net::Server personality
    use Net::LDAP::SimpleServer 'PreFork';

    # using default configuration file
    my $server = Net::LDAP::SimpleServer->new();

    # passing a specific configuration file
    my $server = Net::LDAP::SimpleServer->new({
        conf_file => '/etc/ldapconfig.conf'
    });

    # passing configurations in a hash
    my $server = Net::LDAP::SimpleServer->new({
        port => 5000,
        ldap_data => '/path/to/data.ldif',
    });

    # make it spin
    $server->run();

The default configuration file is:

    ${HOME}/.ldapsimpleserver/config

=head1 DESCRIPTION

=for author to fill in:
    Write a full description of the module and its features here.
    Use subsections (=head2, =head3) as appropriate.

As the name suggests, this module aims to implement a simple LDAP server, 
using many components already available in CPAN. It can be used for
prototyping and/or development purposes. This is B<NOT> intended to be a
production-grade server, altough some brave souls in small offices might
use it as such.

As of November 2010, the server will simply load a LDIF file and serve its
contents through the LDAP protocol. Many operations are B<NOT> available yet,
notably writing into the directory tree, but we would like to implement that
in a near future.


=head1 CONSTRUCTOR 

The constructors will follow the rules defined by L<Net::Server>, but most
notably we have the two forms below:

=over

=item new()

Attempts to create a server by using the default configuration file,
C<< ${HOME}/.ldapsimpleserver/config >>.

=item new( HASHREF )

Attempts to create a server by using the options specified in a hash
reference rather than reading them from a configuration file.

=back

=head1 METHODS 

=over

=item options()

As specified in L<Net::Server>, this method creates new options for the,
server, namely:

=over

ldap_data - the LDIF data file used by LDIFStore

root_dn - the administrator DN of the repository

root_pw - the password for root_dn

=back

=item default_values()

As specified in L<Net::Server>, this method provides default values for a
number of options. In Net::LDAP::SimpleServer, this method is defined as:

    sub default_values {
        return {
            host         => '*',
            port         => 389,
            proto        => 'tcp',
            root_dn      => 'cn=root',
            root_pw      => 'ldappw',
            syslog_ident => 'Net::LDAP::SimpleServer-'
                . $Net::LDAP::SimpleServer::VERSION,
            conf_file => DEFAULT_CONFIG_FILE,
        };
    }

Notice that we do set a default password for the C<< cn=root >> DN. This
allows for out-of-the-box testing, but make sure you change the password
when putting this to production use.

=item post_configure_hook()

Method specified by L<Net::Server> to validate the passed options

=item process_request()

Method specified by L<Net::Server> to actually handle one connection. In this
module it basically delegates the processing to
L<Net::LDAP::SimpleServer::ProtocolHandler>.

=back

=head1 DIAGNOSTICS

=for author to fill in:
    List every single error and warning message that the module can
    generate (even the ones that will "never happen"), with a full
    explanation of each problem, one or more likely causes, and any
    suggested remedies.

=over

=item C<< Error message here, perhaps with %s placeholders >>

[Description of error here]

=item C<< Another error message here >>

[Description of error here]

[Et cetera, et cetera]

=back


=head1 CONFIGURATION AND ENVIRONMENT

=for author to fill in:
    A full explanation of any configuration system(s) used by the
    module, including the names and locations of any configuration
    files, and the meaning of any environment variables or properties
    that can be set. These descriptions must also include details of any
    configuration language used.
  
Net::LDAP::SimpleServer may use a configuration file to specify the
server settings. If no file is specified and options are not passed
in a hash, this module will look for a default configuration file named
C<< ${HOME}/.ldapsimpleserver/config >>. 

    ldap_data /path/to/a/ldif/file.ldif
    #port 389
    #root_dn cn=root
    #root_pw somepassword
    #objectclass_req (true|false)
    #user_tree dc=some,dc=subtree,dc=com
    #user_id_attr uid
    #user_pw_attr password


=head1 DEPENDENCIES

=for author to fill in:
    A list of all the other modules that this module relies upon,
    including any restrictions on versions, and an indication whether
    the module is part of the standard Perl distribution, part of the
    module's distribution, or must be installed separately. ]

L<< Net::LDAP >>

L<< Net::LDAP::Server >>

L<< Net::Server >>

L<< UNIVERSAL::isa >>

L<< Carp >>

L<< File::HomeDir >>

L<< File::Spec::Functions >>

L<< Scalar::Util >>

L<< Config::General >>

L<< Net::LDAP::SimpleServer::LDIFStore >>


=head1 INCOMPATIBILITIES

=for author to fill in:
    A list of any modules that this module cannot be used in conjunction
    with. This may be due to name conflicts in the interface, or
    competition for system or program resources, or due to internal
    limitations of Perl (for example, many modules that use source code
    filters are mutually incompatible).

None reported.


=head1 BUGS AND LIMITATIONS

=for author to fill in:
    A list of known problems with the module, together with some
    indication Whether they are likely to be fixed in an upcoming
    release. Also a list of restrictions on the features the module
    does provide: data types that cannot be handled, performance issues
    and the circumstances in which they may arise, practical
    limitations on the size of data sets, special cases that are not
    (yet) handled, etc.

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-net-ldap-simpleserver@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

Alexei Znamensky  C<< <russoz@cpan.org> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2010, Alexei Znamensky C<< <russoz@cpan.org> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.

